Source/DafnyCore/Compilers/Rust/Dafny-compiler-rust.dfy

```diff
@@ -1,126 +1,508 @@
+      [raw[0]] + ind + AddIndent(raw[1..], ind)
+    else
+      [raw[0]] + AddIndent(raw[1..], ind)
+  }
+
+  datatype Expr =
+      RawExpr(content: string)
+    | Match(matchee: Expr, cases: seq<MatchCase>)
+    | StmtExpr(stmt: Expr, rhs: Expr)
+    | Block(underlying: Expr)
+    | StructBuild(name: string, assignments: seq<AssignIdentifier>)
+  {
+    function ToString(ind: string): string {
+      if Match? then
+        "match " + matchee.ToString(ind + IND) + " {" +
+        SeqToString(cases, (c: MatchCase) requires c < this =>
+                      "\n" + ind + IND + c.ToString(ind + IND), ",") +
+        "\n" + ind + "}"
+      else if StmtExpr? then
+        stmt.ToString(ind) + ";\n" + ind + rhs.ToString(ind)
+      else if Block? then
+        "{\n" + ind + IND + underlying.ToString(ind + IND) + "\n" + ind + "}"
+      else if StructBuild? then
+        name + " {" +
+        SeqToString(assignments, (assignment: AssignIdentifier)
+                    requires assignment < this
+                    =>
+                      "\n" + ind + IND + assignment.ToString(ind + IND), ",") +
+        (if |assignments| > 0 then "\n" + ind else "") + "}"
+      else assert RawExpr?; AddIndent(content, ind)
+    }
+    function Then(rhs2: Expr): Expr {
+      StmtExpr(this, rhs2)
+    }
+  }
+
+  datatype Fn =
+    Fn(name: string, typeParams: seq<TypeParam>, formals: seq<Formal>,
+       returnType: Option<Type>,
+       where: string,
+       body: Option<Expr>)
+  {
+    function ToString(ind: string): string {
+      "fn " + name + TypeParam.ToStringMultiple(typeParams, ind) +
+      "(" + SeqToString(formals, (formal: Formal) => formal.ToString(ind), ", ") + ")" +
+      (match returnType case Some(t) => " -> " + t.ToString(ind) case _ => "") +
+      (if where == "" then "" else "\n" + ind + IND + where) +
+      match body {
+        case None => ";"
+        case Some(body) =>
+          " {\n" + ind + IND +
+          body.ToString(ind + IND) +
+          "\n" + ind + "}"
+      }
+    }
+  }
+}
+
 module {:extern "DCOMP"} DCOMP {
   import opened DAST
+  import Strings = Std.Strings
+  import opened Std.Wrappers
+  import R = RAST
+  const IND := R.IND
+  type Type = DAST.Type
+  type Formal = DAST.Formal
+
+  function runtime(suffix: string): string {
+    "::dafny_runtime" + suffix
+  }
+  const DafnyErasable := runtime("::DafnyErasable")
+
+  // List taken from https://doc.rust-lang.org/book/appendix-01-keywords.html
+  const reserved_rust := {
+    "as","async","await","break","const","continue",
+    "crate","dyn","else","enum","extern","false","fn","for","if","impl",
+    "in","let","loop","match","mod","move","mut","pub","ref","return",
+    "Self","self","static","struct","super","trait","true","type","union",
+    "unsafe","use","where","while","Keywords","The","abstract","become",
+    "box","do","final","macro","override","priv","try","typeof","unsized",
+    "virtual","yield"}
+
+  predicate is_tuple_numeric(i: string) {
```

Comment:
Should we be making all functions opaque by default?

More importantly, it looks like we're not applying any mechanisms to fight brittleness in the Dafny code in DafnyCore. We need to either use the dafny-reportgenerator or set a resource-limit as we did for the standard libraries.

Reply:
Because of the current design, everything verifies quite fast as I'm abiding by the rule that I should not use preconditions or postconditions in functions. Thus, there is little to verify in Dafny and the few lines that Dafny verifies cannot be reduced.
Since there are several competing ways to fight brittleness, I'm making a note to remember that when the time comes. I don't think we need to care about this right now.
What bothers me more is that resolution takes 5 minutes the second I make a new change (and is much faster when I reload the IDE), and verification 20 seconds afterwards.